File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressVerticalGrid/Demos/CBuilder/CalculatedFieldsDemo/CalculatedFieldsDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "CalculatedFieldsDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxDBVGrid"
#pragma link "cxEditRepositoryItems"
#pragma link "cxInplaceContainer"
#pragma link "cxVGrid"
#pragma link "cxButtons"
#pragma link "cxCalendar"
#pragma link "cxContainer"
#pragma link "cxCurrencyEdit"
#pragma link "cxGroupBox"
#pragma link "cxImageList"
#pragma link "cxLabel"
#pragma link "cxMaskEdit"
#pragma link "cxLookAndFeelPainters"
#pragma link "dxGallery"
#pragma link "dxGalleryControl"
#pragma link "dxmdaset"
#pragma link "dxSpreadSheetTypes"
#pragma link "dxSpreadSheetUtils"
#pragma resource "*.dfm"
TfrmMain *frmMain;
//---------------------------------------------------------------------------
__fastcall TfrmMain::TfrmMain(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::FormCreate(TObject *Sender)
{
  String APath = ExtractFilePath(Application->ExeName) + "..\\..\\Data\\";
  cdsProducts2->LoadFromFile(APath + "Products2.cds");
  mdOrder2->LoadFromBinaryFile(APath + "Order2.dat");

  VerticalGrid->DataController->Groups->FullExpand();

  ((TdxGalleryControlOptionsBehaviorAccess*)galColumns->OptionsBehavior)->ItemHotTrack = false;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::btnShowExpressionEditorClick(TObject *Sender)
{
  if (galColumns->Gallery->Groups->Groups[0]->Items->Items[0]->Checked)
	vgrTotal->Properties->ShowExpressionEditor();
  else
	vgrDiscountAmount->Properties->ShowExpressionEditor();
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::ValidateExpressionCell(TcxCustomEditorRowProperties *Sender,
		  int ARecordIndex, const Variant &AValue, TcxEditValidateInfo *AData)

{
  int AErrorCode = VerticalGrid->DataController->ErrorCodes[ARecordIndex][Sender->ItemIndex];
  if (AErrorCode > 0) {
	AData->ErrorType = eetError;
	AData->ErrorText = dxSpreadSheetErrorCodeToString((TdxSpreadSheetFormulaErrorCode)AErrorCode);
  }
}
//---------------------------------------------------------------------------

File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressVerticalGrid/Demos/Delphi/CalculatedFieldsDemo/CalculatedFieldsDemoMain.pas
unit CalculatedFieldsDemoMain;

{$I cxVer.inc}

interface

uses
  SysUtils, Classes, Forms, Controls, Menus, DB, StdCtrls, ComCtrls, DBClient,
  dxCore, cxStyles, cxCustomData, cxGraphics, cxFilter, cxData, cxDataStorage,
  cxEdit, cxClasses, cxControls,
  cxEditRepositoryItems, cxLookAndFeels, cxLookAndFeelPainters,
  cxNavigator, cxContainer, cxGroupBox, ActnList, cxCheckBox, cxLabel,
  cxTextEdit, cxMaskEdit, cxSpinEdit, cxDropDownEdit, cxImageComboBox, cxDBData,
  DemoBasicMain, MidasLib,
  cxDataControllerConditionalFormattingRulesManagerDialog, cxGridLevel,
  cxGridCustomTableView, cxGridTableView, cxGridDBTableView, cxGridCustomView,
  cxGrid, ImgList, cxVGrid, cxDBVGrid, cxInplaceContainer, dxmdaset, dxGallery, dxGalleryControl, cxButtons, cxCalendar,
  cxCurrencyEdit, cxImageList;

type
  TfrmMain = class(TDemoBasicMainForm)
    cxGroupBox1: TcxGroupBox;
    cxGroupBox2: TcxGroupBox;
    btnShowExpressionEditor: TcxButton;
    galColumns: TdxGalleryControl;
    dxGalleryControl1Group1: TdxGalleryControlGroup;
    dxGalleryControl1Group1Item1: TdxGalleryControlItem;
    dxGalleryControl1Group1Item2: TdxGalleryControlItem;
    VerticalGrid: TcxDBVerticalGrid;
    vgrCategoryRow1: TcxCategoryRow;
    vgrOrderID: TcxDBEditorRow;
    vgrOrderDate: TcxDBEditorRow;
    vgrCategoryRow2: TcxCategoryRow;
    vgrProductName: TcxDBEditorRow;
    vgrUnitPrice: TcxDBEditorRow;
    vgrQuantity: TcxDBEditorRow;
    vgrDiscount: TcxDBEditorRow;
    vgrCategoryRow3: TcxCategoryRow;
    vgrDiscountAmount: TcxDBEditorRow;
    vgrTotal: TcxDBEditorRow;
    cdsProducts2: TClientDataSet;
    cdsProducts2ProductID: TIntegerField;
    cdsProducts2ProductName: TStringField;
    cdsProducts2SupplierID: TIntegerField;
    cdsProducts2CategoryID: TIntegerField;
    cdsProducts2QuantityPerUnit: TStringField;
    cdsProducts2UnitPrice: TFloatField;
    cdsProducts2UnitsInStock: TIntegerField;
    cdsProducts2UnitsOnOrder: TIntegerField;
    cdsProducts2ReorderLevel: TIntegerField;
    cdsProducts2Discontinued: TBooleanField;
    cdsProducts2EAN13: TStringField;
    mdOrder2: TdxMemData;
    mdOrder2OrderID: TIntegerField;
    mdOrder2ProductID: TIntegerField;
    mdOrder2UnitPrice: TFloatField;
    mdOrder2Quantity: TIntegerField;
    mdOrder2Discount: TFloatField;
    mdOrder2OrderDate: TDateTimeField;
    mdOrder2ProductName: TStringField;
    dsOrder2: TDataSource;
    cxImageList1: TcxImageList;
    cxLabel1: TcxLabel;
    procedure FormCreate(Sender: TObject);
    procedure btnShowExpressionEditorClick(Sender: TObject);
    procedure ValidateExpressionCell(Sender: TcxCustomEditorRowProperties;
      ARecordIndex: Integer; const AValue: Variant; AData: TcxEditValidateInfo);
  end;

var
  frmMain: TfrmMain;

implementation

{$R *.dfm}

uses
  dxSpreadSheetTypes, dxSpreadSheetUtils;

type
  TdxGalleryControlOptionsBehaviorAccess = class(TdxGalleryControlOptionsBehavior);

procedure TfrmMain.btnShowExpressionEditorClick(Sender: TObject);
begin
  if galColumns.Gallery.Groups[0].Items[0].Checked then
    vgrTotal.Properties.ShowExpressionEditor
  else
    vgrDiscountAmount.Properties.ShowExpressionEditor;
end;

procedure TfrmMain.FormCreate(Sender: TObject);
var
  APath: string;
begin
  APath := ExtractFilePath(Application.ExeName) + '..\..\Data\';
  cdsProducts2.LoadFromFile(APath + 'Products2.cds');
  mdOrder2.LoadFromBinaryFile(APath + 'Order2.dat');

  TdxGalleryControlOptionsBehaviorAccess(galColumns.OptionsBehavior).ItemHotTrack := False;
end;

procedure TfrmMain.ValidateExpressionCell(Sender: TcxCustomEditorRowProperties;
  ARecordIndex: Integer; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorCode: Integer;
begin
  AErrorCode := VerticalGrid.DataController.ErrorCodes[ARecordIndex, Sender.ItemIndex];
  if AErrorCode > 0 then
  begin
    AData.ErrorType := eetError;
    AData.ErrorText := dxSpreadSheetErrorCodeToString(TdxSpreadSheetFormulaErrorCode(AErrorCode));
  end;
end;

end.
